presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java

```diff
@@ -238,6 +268,36 @@ else if (controlResult.getState() != State.SUCCESS) {
         return true;
     }
 
+    private QueryResult executeQueryTest()
+    {
+        Query query = queryPair.getTest();
+        QueryResult setupResult = setup(query, testPrequery -> executeQuery(testGateway, testUsername, testPassword, queryPair.getTest(), testPrequery, testTimeout, sessionProperties));
+        if (setupResult.getState() != State.SUCCESS) {
+            return setupResult;
+        }
+        QueryResult queryResult = executeQuery(testGateway, testUsername, testPassword, queryPair.getTest(), query.getQuery(), testTimeout, sessionProperties);
+        if (queryResult.getState() != State.SUCCESS) {
+            return queryResult;
+        }
+        QueryResult tearDownResult = tearDown(query, testPostquery -> executeQuery(testGateway, testUsername, testPassword, queryPair.getTest(), testPostquery, testTimeout, sessionProperties));
+        return tearDownResult.getState() == State.SUCCESS ? queryResult : tearDownResult;
+    }
+
+    private QueryResult executeQueryControl()
```

Comment:
this won't run the teardown queries if the setup or test query fails, or if there's an exception. Put the tear down in a finally-block